tensorflow/lite/micro/tools/python_interpreter/src/interpreter_wrapper.cc

```diff
@@ -0,0 +1,227 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "tensorflow/lite/micro/tools/python_interpreter/src/interpreter_wrapper.h"
+
+// Disallow Numpy 1.7 deprecated symbols.
+#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION
+// See https://numpy.org/doc/1.16/reference/c-api.array.html#importing-the-api
+#define NO_IMPORT_ARRAY
+#define PY_ARRAY_UNIQUE_SYMBOL tflite_micro_python_interpreter_array_api
+#include <numpy/arrayobject.h>
+#include <pybind11/pybind11.h>
+
+#include "tensorflow/lite/c/common.h"
+#include "tensorflow/lite/core/api/error_reporter.h"
+#include "tensorflow/lite/kernels/kernel_util.h"
+#include "tensorflow/lite/micro/all_ops_resolver.h"
+#include "tensorflow/lite/micro/micro_error_reporter.h"
+#include "tensorflow/lite/micro/micro_interpreter.h"
+#include "tensorflow/lite/micro/tools/python_interpreter/src/numpy_utils.h"
+#include "tensorflow/lite/micro/tools/python_interpreter/src/python_utils.h"
+
+namespace py = pybind11;
+
+namespace tflite {
+
+InterpreterWrapper::~InterpreterWrapper() {
+  // Undo any references incremented
+  Py_DECREF(model_);
+}
+
+InterpreterWrapper::InterpreterWrapper(PyObject* model_data,
+                                       size_t arena_size) {
+  // `model_data` is used as a raw pointer beyond the scope of this
+  // constructor, so we need to increment the reference count so that Python
+  // doesn't destroy it during the lifetime of this interpreter.
+  Py_INCREF(model_data);
+
+  // Get the input array contained in `model_data` as a byte array
+  char* buf = nullptr;
+  Py_ssize_t length;
+  if (ConvertFromPyString(model_data, &buf, &length) == -1 || buf == nullptr) {
+    PyErr_SetString(
+        PyExc_ValueError,
+        "TFLM cannot convert model data from Python object to char *");
+    return;
+  }
+
+  const Model* model = GetModel(buf);
+  std::unique_ptr<ErrorReporter> error_reporter(new MicroErrorReporter());
+  std::unique_ptr<uint8_t[]> tensor_arena(new uint8_t[arena_size]);
+  std::unique_ptr<MicroInterpreter> interpreter(
+      new MicroInterpreter(model, all_ops_resolver_, tensor_arena.get(),
+                           arena_size, error_reporter.get()));
+
+  // Save variables that need to be used or destroyed later
+  model_ = model_data;
+  error_reporter_ = std::move(error_reporter);
+  interpreter_ = std::move(interpreter);
+  memory_arena_ = std::move(tensor_arena);
+
+  TfLiteStatus status = interpreter_->AllocateTensors();
+  if (status != kTfLiteOk) {
+    PyErr_SetString(PyExc_RuntimeError, "TFLM failed to allocate tensors");
+    return;
+  }
+
+  // This must be called before using any PyArray_* APIs. It essentially sets
```

Comment:
does this need to be the final statement or could this at the top as part of setup?

Reply:
The order doesn't matter as much here because we're not using any PyArray_* API in the constructor. We just need to do the import anytime before calling any PyArray_* function.